lib/AST/NameLookup.cpp

```diff
@@ -1483,6 +1505,55 @@ populateLookupTableEntryFromExtensions(ASTContext &ctx,
   }
 }
 
+static void
+populateLookupTableEntryFromMacroExpansions(ASTContext &ctx,
+                                            MemberLookupTable &table,
+                                            DeclName name,
+                                            NominalTypeDecl *dc) {
+  auto expandAndPopulate = [&](MacroExpansionDecl *med) {
+    auto expanded = evaluateOrDefault(med->getASTContext().evaluator,
+                                      ExpandMacroExpansionDeclRequest{med},
+                                      nullptr);
+    for (auto *decl : expanded)
+      table.addMember(decl);
+  };
+
+  for (auto *member : dc->getCurrentMembersWithoutLoading()) {
+    auto *med = dyn_cast<MacroExpansionDecl>(member);
+    if (!med)
+      continue;
+    auto macro = evaluateOrDefault(
+        ctx.evaluator, ResolveMacroRequest{med, MacroRole::Declaration, dc},
+        nullptr);
+    if (!macro)
+      continue;
+    auto *attr = macro->getMacroRoleAttr(MacroRole::Declaration);
+    // If a macro produces arbitrary names, we have to expand it to factor its
+    // expansion results into name lookup.
+    if (attr->hasNameKind(MacroIntroducedDeclNameKind::Arbitrary)) {
+      expandAndPopulate(med);
+    }
+    // Otherwise, we expand the macro if it has the same decl base name being
+    // looked for.
+    else {
+      auto it = llvm::find_if(attr->getNames(),
+                              [&](const MacroIntroducedDeclName &introName) {
+        // FIXME: The `Named` kind of `MacroIntroducedDeclName` should store a
```

Comment:
There's more to check here (prefix/suffix, etc.) but the structure here looks great.

Reply:
Yeah, I didn't check for those here because for `MacroExpansionDecl` in a type context, only `named` and `arbitrary` are allowed.